### PR TITLE
Autopilot: Change the types to the domain type

### PR DESCRIPTION
**Pull request by Mateo-mro**
# Description
It seems we were not using the domain types in the domain dtos. Changing the dtos to use the domain types in the autopilot.

# Changes
- Changed the dtos to use the domain types in the autopilot.

## How to test
1. Regression tests

### CODE REVIEW

File: crates/autopilot/src/domain/quote/mod.rs [hunk 1]
```diff
@@ -1,9 +1,9 @@
...
 
 #[derive(Clone, Debug)]
 pub struct Quote {
     pub order_uid: OrderUid,
-    pub sell_amount: U256,
-    pub buy_amount: U256,
-    pub fee: U256,
+    pub sell_amount: eth::TokenAmount,
+    pub buy_amount: eth::TokenAmount,
+    pub fee: eth::TokenAmount,
```
Comment (sunce86) [inline crates/autopilot/src/domain/quote/mod.rs:8]: Would be really nice for this to be `SellTokenAmount`

File: crates/autopilot/src/domain/settlement/transaction.rs [hunk 2]
```diff
@@ -12,7 +12,7 @@ pub struct Transaction {
     /// The block number of the block that contains the transaction.
     pub block: eth::BlockNo,
     /// The gas used by the transaction.
-    pub gas: eth::U256,
+    pub gas: eth::TokenAmount,
```
Comment (sunce86) [inline crates/autopilot/src/domain/settlement/transaction.rs:15]: not actually a token

File: crates/autopilot/src/domain/settlement/transaction.rs [hunk 3]
```diff
@@ -12,7 +12,7 @@ pub struct Transaction {
...
     /// The effective gas price of the transaction.
-    pub effective_gas_price: eth::U256,
+    pub effective_gas_price: eth::TokenAmount,
```
Comment (sunce86) [inline crates/autopilot/src/domain/settlement/transaction.rs:17]: same as ☝️

**Review by sunce86 — COMMENTED** [review]
Good direction.

I did plan to do this, however there are some suggestions I made that are more involving change so that was the reason I did not have time to tackle this.

Still fine to merge in this state if others like it, while we can make further improvements down the road.

File: crates/autopilot/src/domain/auction/order.rs [hunk 4]
```diff
@@ -7,18 +7,18 @@ use {
 #[derive(Clone, Debug, PartialEq)]
 pub struct Order {
     pub uid: OrderUid,
-    pub sell_token: H160,
-    pub buy_token: H160,
-    pub sell_amount: U256,
-    pub buy_amount: U256,
+    pub sell_token: eth::TokenAddress,
+    pub buy_token: eth::TokenAddress,
```
[the remark this answers is not in the document]
Comment (Mateo-mro) [reply crates/autopilot/src/domain/auction/order.rs:11]: @sunce86 for my understanding, what is the difference between them? I know `Asset` is a wider concept, so we plan to change everything to `Asset` in the future?

File: crates/autopilot/src/domain/settlement/transaction.rs (hunk 2 above)
Comment (MartinquaXD) [reply crates/autopilot/src/domain/settlement/transaction.rs:15]: Why was this resolved? It's still marked as a token amount.

Comment (Mateo-mro) [reply crates/autopilot/src/domain/settlement/transaction.rs:15]: @MartinquaXD I resolve comments as I do to keep track, and then I push it.

File: crates/autopilot/src/infra/persistence/dto/auction.rs [hunk 5]
```diff
@@ -30,7 +34,17 @@ pub fn to_domain(auction: Auction) -> domain::Auction {
             .into_iter()
             .map(super::order::to_domain)
             .collect(),
-        prices: auction.prices,
+        prices: auction
+            .prices
+            .into_iter()
+            .filter_map(|(key, value)| {
+                let price = Price::new(value.into());
+                if let Err(ref e) = price {
```
Comment (Mateo-mro) [inline crates/autopilot/src/infra/persistence/dto/auction.rs:42]: Should I error silently here? or should I change this function from `to_domain()` to `try_to_domain()` and return a `Result<>`?

File: crates/autopilot/src/domain/eth/mod.rs [hunk 6]
```diff
@@ -45,6 +45,19 @@ impl TokenAmount {
     }
 }
 
+/// An ERC20 sell token amount.
+///
+/// https://eips.ethereum.org/EIPS/eip-20
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, From, Into)]
```
Comment (Mateo-mro) [inline crates/autopilot/src/domain/eth/mod.rs:51]: shouldn't we have a crate containing all these types? they seem to repeat themselves over the codebase :thinking:

File: crates/autopilot/src/domain/settlement/transaction.rs (hunk 2 above)
Comment (squadgazzz) [reply crates/autopilot/src/domain/settlement/transaction.rs]: > I resolve comments as I do locally to keep track, and then I push it.

Not really an intuitive approach from the reviewer's perspective 🙂

File: crates/autopilot/src/domain/eth/mod.rs (hunk 6 above)
Comment (squadgazzz) [reply crates/autopilot/src/domain/eth/mod.rs:51]: Would the `model` crate be an appropriate place for those?

File: crates/autopilot/src/infra/persistence/dto/auction.rs (hunk 5 above)
Comment (squadgazzz) [reply crates/autopilot/src/infra/persistence/dto/auction.rs:47]: This gets called from the [**function**](https://github.com/cowprotocol/services/blob/bc2ebebfb64eeb62c9219977b595058c5e4dd7cb/crates/autopilot/src/infra/shadow/orderbook/mod.rs#L28) returning `Result` so it might make sense to return a `Result` here as well.

File: crates/autopilot/src/domain/settlement/transaction.rs (hunk 2 above)
Comment (Mateo-mro) [reply crates/autopilot/src/domain/settlement/transaction.rs]: > Not really an intuitive approach from the reviewer's perspective 🙂

What I can do is to put the PR in draft while reworking it. If that helps. But usually the time between me resolving a conversion and then pushing the code is not more than 10-15 minutes.

File: crates/autopilot/src/domain/eth/mod.rs (hunk 6 above)
Comment (Mateo-mro) [reply crates/autopilot/src/domain/eth/mod.rs:51]: No, those types are not models, they are domain types. I wouldn't place them in the `model` crate, since `model` implies the types implement `serde` (which is not the case)

File: crates/autopilot/src/infra/persistence/dto/auction.rs (hunk 5 above)
Comment (Mateo-mro) [reply crates/autopilot/src/infra/persistence/dto/auction.rs:47]: I also like more the approach of returning an `Result<>` (it is cleaner).

File: crates/autopilot/src/domain/eth/mod.rs (hunk 6 above)
Comment (fleupold) [reply crates/autopilot/src/domain/eth/mod.rs:51]: Would you want to share these types across driver and autopilot? I'm not sure those should have a coupled domain (it's not so much work to re-define them and I'm worried we will start leaking things that would be driver specific into a shared crate).

Although there is an argument to be made that their may be some over-arching COW domain (which includes types around tokens, trades, etc). But frankly this is how the model crate came along which has matured pretty poorly imo.

File: crates/autopilot/src/infra/persistence/dto/auction.rs (hunk 5 above)
Comment (fleupold) [reply crates/autopilot/src/infra/persistence/dto/auction.rs:47]: +1 for returning this error instead of creating a potentially inconsistent auction (it doesn't seem as if we can guarantee all prices are non-zero).

**Review by fleupold — APPROVED** [review]

File: crates/autopilot/src/domain/eth/mod.rs (hunk 6 above)
Comment (Mateo-mro) [reply crates/autopilot/src/domain/eth/mod.rs:51]: @fleupold (also cc @sunce86 for feedback), my idea is to have a crate called `cow-types` which contains very basic type which are 100 % common to every domain. For example:

1. type `Address` defined exactly the same in three different places:
- [Autopilot](https://github.com/cowprotocol/services/blob/main/crates/autopilot/src/domain/eth/mod.rs#L6)
- [Driver](https://github.com/cowprotocol/services/blob/main/crates/driver/src/domain/eth/mod.rs#L111)
- [Solvers](https://github.com/cowprotocol/services/blob/main/crates/solvers/src/domain/eth/mod.rs#L90)

Basically majority of the `eth` module is copied all over at least those three crates. The idea of `cow-types` is to place "absolute truth" types, such as `Asset`, `SellTokenAmount`, `Token`, etc. I don't want/think placing in such crate domain types such as `Auction` is a good idea, because `Auction` is no absolute truth for every domain, since each domain may have their own implementation/definition of it.

Comment (Mateo-mro) [reply crates/autopilot/src/domain/eth/mod.rs:51]: I would like to also make those types not to leak their internal type (e.g. in case of the `Address` type, its `H160` type). Derive `From/Into` with the `derive_more` crate I introduced into the workspace crate, and do `H160::from()` or `variable.into()` for getting its `H160` instead of `.0`. So it would behave like a "proper type crate".

File: crates/autopilot/src/infra/shadow/orderbook/mod.rs [hunk 7]
```diff
@@ -17,14 +17,15 @@ impl Orderbook {
     }
 
     /// Retrieves the current auction.
-    pub async fn auction(&self) -> reqwest::Result<domain::AuctionWithId> {
+    pub async fn auction(&self) -> anyhow::Result<domain::AuctionWithId> {
```
Comment (Mateo-mro) [inline crates/autopilot/src/infra/shadow/orderbook/mod.rs:20]: The type of error is not important for the caller, so it is better to transform from `reqwest::Error` error to `anyhow::Error` than the other way around.

File: crates/autopilot/src/infra/persistence/dto/auction.rs [hunk 8]
```diff
@@ -26,22 +33,29 @@ pub fn from_domain(auction: domain::Auction) -> Auction {
...
             .collect(),
-        prices: auction.prices,
+        prices: {
+            let mut result = Prices::new();
+            auction.prices.into_iter().try_for_each(|(key, value)| {
+                result.insert(key.into(), Price::new(value.into())?);
+                Ok::<_, anyhow::Error>(())
+            })?;
+            result
+        },
```
Comment (sunce86) [inline crates/autopilot/src/infra/persistence/dto/auction.rs:52]: nit: This should also work i think, so whatever you find prettier

```
prices: auction
            .prices
            .into_iter()
            .map(|(key, value)| {
                Price::new(value.into()).map(|price| (domain::eth::TokenAddress(key), price))
            })
            .collect::<Result<_, _>>()?,
```

**Review by sunce86 — COMMENTED** [review]
Lg, mostly nits just a small remark around GasPrice

**Review by sunce86 — APPROVED** [review]

**Review by MartinquaXD — APPROVED** [review]

File: crates/autopilot/src/domain/eth/mod.rs (hunk 6 above)
Comment (sunce86) [reply crates/autopilot/src/domain/eth/mod.rs:51]: I'd prefer to keep them separated. They are part of their `domain` modules and probably shouldn't be shared across multiple crates. We can't guarantee they will always stay same for all crates.
